api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java

```diff
@@ -387,18 +406,181 @@ public boolean supportsPropertyName(String gpName) {
+				return null;
+			} else {
+				Patient patient = patients.get(0);
+				dao.getSessionFactory().getCurrentSession().refresh(patient);
+				return patient;
+			}
+			//TODO Add error if more than one patient is has that identifier!
 		}
 	}
 
+	private Patient persistImportedMpiPatient(MpiPatient mpiPatient) {
+		// TODO: Fix bug MPI-11 Updating of Locally generated ID in MPI after MPI import not occurring
+		String openMrsIdTypeUuid = adminService.getGlobalProperty(RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_UUID);
+
+		PatientIdentifierType openMrsIdType = patientService.getPatientIdentifierTypeByUuid(openMrsIdTypeUuid);
+		Patient patient;
+		if (mpiPatient.getPatientIdentifier(openMrsIdType) == null) {
+			PatientIdentifier localId = validateOrGenerateIdentifier(null, null);
+			mpiPatient.addIdentifier(localId);
+		}
+		patient = patientService.savePatient(mpiPatient);
+		return patient;
+	}
+
+	private IdentifierSourceService getIssAndUpdateIdSource() {
+		IdentifierSourceService iss = Context.getService(IdentifierSourceService.class);
+		if (idSource == null) {
+			String idSourceId = adminService.getGlobalProperty(RegistrationCoreConstants.GP_OPENMRS_IDENTIFIER_SOURCE_ID);
+			if (StringUtils.isBlank(idSourceId)) {
+				throw new APIException("Please set the id of the identifier source to use to generate patient identifiers");
+			}
+			try {
+				idSource = iss.getIdentifierSource(Integer.valueOf(idSourceId));
+				if (idSource == null) {
+					throw new APIException("cannot find identifier source with id:" + idSourceId);
+				}
+			}
+			catch (NumberFormatException e) {
+				throw new APIException("Identifier source id should be a number");
+			}
+		}
+
+		return iss;
+	}
+
+	private Location getIdentifierLocation(Location identifierLocation) {
+		if (identifierLocation == null) {
+			identifierLocation = locationService.getDefaultLocation();
+			if (identifierLocation == null) {
+				throw new APIException("Failed to resolve location to associate to patient identifiers");
+			}
+		}
+
+		// see if there is a primary identifier location further up the chain, use that instead if there is
+		Location identifierAssignmentLocationAssociatedWith = getIdentifierAssignmentLocationAssociatedWith(identifierLocation);
+		if (identifierAssignmentLocationAssociatedWith != null) {
+			identifierLocation = identifierAssignmentLocationAssociatedWith;
+		}
+
+		return identifierLocation;
+	}
+
+	private PatientIdentifier validateOrGenerateIdentifier(String identifierString, Location identifierLocation) {
+		IdentifierSourceService iss = getIssAndUpdateIdSource();
+		identifierLocation = getIdentifierLocation(identifierLocation);
+
+		// generate identifier if necessary, otherwise validate
+		if (StringUtils.isBlank(identifierString)) {
+			identifierString = iss.generateIdentifier(idSource, null);
+		} else {
+			PatientIdentifierValidator.validateIdentifier(identifierString, idSource.getIdentifierType());
```

Comment:
I don't think you need to validate the Id here because the API will do it anyways when you call savePatient

Reply:
@mogoodrich this validation line was added by you [here](https://github.com/openmrs/openmrs-module-registrationcore/blob/e7d2d6b773adc462899b7c1bccac87f0e69aa321/api/src/main/java/org/openmrs/module/registrationcore/api/impl/RegistrationCoreServiceImpl.java#L201) in [this commit](https://github.com/openmrs/openmrs-module-registrationcore/commit/7b0a3224434bbcd4ccfa3efda6bba8153eba1177) Any reason to double check the id?